eliot/_action.py

```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
+    level.
+    """
+
+    def __init__(self, action, message):
+        Exception.__init__(
+            self, 'Tried to add {} to {}, but already had child at {}'.format(
+                message, action, message.task_level))
+
+
+class InvalidStartMessage(Exception):
+    """
+    Tried to start an action with an invalid message.
+    """
+
+    def __init__(self, message, reason):
+        Exception.__init__(
+            self, 'Invalid start message {}: {}'.format(message, reason))
+
+    @classmethod
+    def wrong_status(cls, message):
+        return cls(message, 'must have status "STARTED"')
+
+    @classmethod
+    def wrong_task_level(cls, message):
+        return cls(message, 'first message must have task level ending in 1')
+
+
+class WrittenAction(PClass):
+    """
+    An Action that has been logged.
+
+    @ivar action_type: The type of the action,
+        e.g. C{"yourapp:subsystem:dosomething"}
+
+    @ivar status: One of C{STARTED_STATUS}, C{SUCCEEDED_STATUS}, or
+        C{FAILED_STATUS}.
+
+    @ivar task_uuid: The UUID of the task to which this action belongs.
+
+    @ivar TaskLevel task_level: The level of the task in which the action occurs.
+
+    @ivar start_time: The Unix timestamp of when the action started.
+
+    @ivar end_time: The Unix timestamp of when the action ended, or C{None} if
+        there has been no end message.
+
+    @ivar exception: If the action failed, the name of the exception that was
+        raised to cause it to fail. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+
+    @ivar reason: The reason the action failed. If the action succeeded, or
+        hasn't finished yet, then C{None}.
+
+    @ivar _children: A L{pvector} of L{WrittenAction} and L{WrittenMessage}
+        that make up this action.
+    """
+
+    # Broad philosophy:
+    # - Have a way of constructing known-valid instances (at the very least,
+    #   this is useful for tests)
+    # - Have a thing that takes messages out of order and hangs on to them
+    #   until it can construct known-valid instances
+    # - Be able to recover the original messages
+
+    action_type = field(type=unicode, mandatory=True)
+    status = field(type=unicode, mandatory=True,
+                   invariant=lambda x: (x in VALID_STATUSES, 'Invalid status'))
+    task_uuid = field(type=UUID, mandatory=True)
+    task_level = field(type=TaskLevel, mandatory=True)
+    start_time = field(type=float, mandatory=True)
+    end_time = field(type=optional(float), mandatory=True)
+    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
+    # pyrsistent doesn't support recursive data types:
+    # https://github.com/tobgu/pyrsistent/issues/48
+    _children = pmap_field(TaskLevel, object)
+    exception = field(type=optional(unicode), mandatory=True)
+    reason = field(type=optional(unicode), mandatory=True)
+
+    # XXX: Possible invariants:
+    # - start_time <= end_time
```

Comment:
Quite. I think it would be an interesting thing for a log display tool to highlight.